handler/src/main/java/io/netty/handler/ssl/ConscryptAlpnSslEngine.java

```diff
@@ -0,0 +1,239 @@
+    }
+
+    @Override
+    public final SSLEngineResult wrap(ByteBuffer[] byteBuffers, int i, int i2, ByteBuffer byteBuffer)
+            throws SSLException {
+        return updateHandshakeStatus(super.wrap(byteBuffers, i, i2, byteBuffer));
+    }
+
+    @Override
+    public final SSLEngineResult unwrap(ByteBuffer byteBuffer, ByteBuffer byteBuffer2) throws SSLException {
+        return updateHandshakeStatus(super.unwrap(byteBuffer, byteBuffer2));
+    }
+
+    @Override
+    public final SSLEngineResult unwrap(ByteBuffer byteBuffer, ByteBuffer[] byteBuffers) throws SSLException {
+        return updateHandshakeStatus(super.unwrap(byteBuffer, byteBuffers));
+    }
+
+    @Override
+    public final SSLEngineResult unwrap(ByteBuffer byteBuffer, ByteBuffer[] byteBuffers, int i, int i2)
+            throws SSLException {
+        return updateHandshakeStatus(super.unwrap(byteBuffer, byteBuffers, i, i2));
+    }
+
+    abstract void onHandshakeComplete() throws SSLException;
+
+    private SSLEngineResult updateHandshakeStatus(SSLEngineResult result) throws SSLException {
+        if (!handshakeComplete && isHandshakeComplete(result.getHandshakeStatus())) {
+            // Notify handshake completion even though handshaking is being skipped.
+            handshakeComplete = true;
+            onHandshakeComplete();
+        }
+        return result;
+    }
+
+    private void updateHandshakeStatus() throws SSLException {
+        if (!handshakeComplete && isHandshakeComplete(getHandshakeStatus())) {
+            // Notify handshake completion even though handshaking is being skipped.
+            handshakeComplete = true;
+            onHandshakeComplete();
+        }
+    }
+
+    private static boolean isHandshakeComplete(HandshakeStatus handshakeStatus) {
+        switch (handshakeStatus) {
+            case NOT_HANDSHAKING:
+            case FINISHED:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    private ConscryptAlpnSslEngine(OpenSSLEngineImpl engine, List<String> protocols) {
+        super(engine);
+
+        // Set the list of supported ALPN protocols on the engine.
+        setAlpnProtocols(engine, protocols);
+    }
+
+    OpenSSLEngineImpl getWrappedConscryptEngine() {
+        return (OpenSSLEngineImpl) getWrappedEngine();
+    }
+
+    private static void setAlpnProtocols(OpenSSLEngineImpl engine, List<String> protocols) {
+        engine.setAlpnProtocols(toProtocolArray(protocols));
+    }
+
+    private static String getAlpnSelectedProtocol(OpenSSLEngineImpl engine) {
+        byte[] data = engine.getAlpnSelectedProtocol();
+        try {
+            return data == null ? null : new String(data, CharsetUtil.US_ASCII.name());
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static byte[] toProtocolArray(List<String> protocols) {
```

Comment:
Agreed. I think a better place for this would be in Conscrypt, itself.  Let me look into that.